handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -917,16 +927,19 @@ public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
         discardSomeReadBytes();
 
         flushIfNeeded(ctx);
+        readIfNeeded(ctx);
+
+        firedChannelRead = false;
+        ctx.fireChannelReadComplete();
+    }
 
+    private void readIfNeeded(ChannelHandlerContext ctx) {
         // If handshake is not finished yet, we need more data.
         if (!ctx.channel().config().isAutoRead() && (!firedChannelRead || !handshakePromise.isDone())) {
             // No auto-read used and no message passed through the ChannelPipeline or the handhshake was not complete
             // yet, which means we need to trigger the read to ensure we not encounter any stalls.
             ctx.read();
```

Comment:
Right. I think that's what we've been doing before. My take it that a read initiated by `SslHanlder` to finish the handshake shouldn't affect the `readDuringHandshake` flag. Although, I can see some sense from the opposite relation. Maybe we don't need to trigger a `read()` if `readDuringHandshake` is true. But again, we don't check that today.